File: SBiswas/Problems/Milestone1/Problem16.cpp
/* Average of levels of a binary tree*/

#include <iostream>
#include <vector>
#include <queue>


using namespace std;

class TreeNode {
    
    public:
    int data;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
    
};

class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {

        queue<TreeNode*> curr_level;
        queue<TreeNode*> next_level;

        vector<double> ans;

        next_level.push(root);
        ans.push_back((double)root->data);


        while(!next_level.empty())
        {
            curr_level = next_level;
            queue<TreeNode*> temp;
            next_level = temp;

            while(!curr_level.empty())
            {
                TreeNode* top = curr_level.front();
                curr_level.pop();

                if(top->left!=nullptr)
                {
                    next_level.push(top->left);
                }

                if(top->right!=nullptr)
                {
                    next_level.push(top->right);
                }
            }

            double sum = 0;
            double count=0;

            if(!next_level.empty())
            {
                temp = next_level;
                while(!temp.empty())
                {
                    sum+= (double)temp.front()->data;
                    count++;
                    temp.pop();
                }
                ans.push_back(sum/count);
            }

        }

        return ans;
        
    }
};

File: Hritwik/Problems/Milestone1/Problem16.py
# Definition for a binary tree node.
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def averageOfLevels(self, root):
        if not root:
            return []
        queue = deque([root])
        response = []
        while queue:
            len_of_queue = len(queue)
            sum=0
            level = []
            for i in range(len_of_queue):
                node = queue.popleft()
                level.append(node.val)
                sum+=node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            avg_of_level = sum/len_of_queue
            response.append(avg_of_level)
        return response


solution = Solution()

# Create a binary tree to traverse through it in an inorder traversal manner
#
#         Root   2
#              /    \
#             2      5  
#                   /  \
#                  5    7
#
root = TreeNode(2)
root.left = TreeNode(2)
root.right = TreeNode(5)
root.right.left = TreeNode(5)
root.right.right = TreeNode(7)


response = solution.averageOfLevels(root)
print(f'The average value of the nodes on each level in the form of an array. -> {response}')
